### PR TITLE
Fix: `external_data_configuration.connection_id` Diff Suppression

### PR DESCRIPTION
**Pull request by nevzheng**
Fixes: https://github.com/hashicorp/terraform-provider-google/issues/12386

Mentioned in; https://github.com/hashicorp/terraform-provider-google/issues/12465

# Changes
+ Add: Test Coverage for `name` and `id` references in External Data Configuration
+ Fix: case sensitivity in location for connection id. In the slash and dot forms of connection id the region has different case. Change the comparison to be case insensittive

**Release Note Template for Downstream PRs (will be copied)**

```release-note:bug
bigquery: fix diff suppression in `external_data_configuration.connection_id`
```

### CODE REVIEW

**Comment by nevzheng** [comment]
Demonstration of the Issue I'm Seeing.

image
image

File: mmv1/third_party/terraform/services/bigquery/resource_bigquery_table.go [hunk 1]
```diff
@@ -191,7 +191,11 @@ func bigQueryTableConnectionIdSuppress(name, old, new string, _ *schema.Resource
 
 	re := regexp.MustCompile("projects/(.+)/(?:locations|regions)/(.+)/connections/(.+)")
 	if matches := re.FindStringSubmatch(new); matches != nil {
-		return old == matches[1]+"."+matches[2]+"."+matches[3]
+		// The Location or Region may be returned in Upper or Lower Case.
+		normalizedIdWithUpperCaseLocation := matches[1] + "." + strings.ToUpper(matches[2]) + "." + matches[3]
+		normalizedIdWithLowerCaseLocation := matches[1] + "." + strings.ToLower(matches[2]) + "." + matches[3]
+
+		return (old == normalizedIdWithUpperCaseLocation) || (old == normalizedIdWithLowerCaseLocation)
```
Comment (ScottSuarez) [inline mmv1/third_party/terraform/services/bigquery/resource_bigquery_table.go:198]: Are we sure there are no mixed case scenarios?

Would it make more sense to normalize both old and new region/zone then do the comparison?

Comment (nevzheng) [reply mmv1/third_party/terraform/services/bigquery/resource_bigquery_table.go:198]: IIUC It's not safe to do a case insensititve comparison accross the whole string. project and location i think we can do a case insensitive comparison, but i'm not sure about connection IDs.

# Project Strings
When I open the cloud console and try to create a new project, I see the following restriction. 
```
Project ID can have lowercase letters, digits, or hyphens. It must start with a lowercase letter and end with a letter or number.
```

# Region and Location Strings
IIUC, Regions and location strings are case insenstive. they're either all upper or lower case and may include digits and `-` (dashes).

https://cloud.google.com/storage/docs/locations#available-locations
https://cloud.google.com/bigquery/docs/locations#multi-regions

## Connection ID names
I used the API explorer and was able to create connection IDs with mixed case.

https://cloud.google.com/bigquery/docs/reference/bigqueryconnection/rest/v1/projects.locations.connections/create

Additionally, It looks like the API always returns the lower case version of the region in the dot string, so lets just force the returned location in slash to be lower case.

Comment (ScottSuarez) [reply mmv1/third_party/terraform/services/bigquery/resource_bigquery_table.go]: apologies what I was inferring was splitting the old and new and setting region/zone to lowercase and then remerging. Similar to what you are doing on the new

Comment (ScottSuarez) [reply mmv1/third_party/terraform/services/bigquery/resource_bigquery_table.go]: Not normalizing the entire string

Comment (nevzheng) [reply mmv1/third_party/terraform/services/bigquery/resource_bigquery_table.go]: done.

i parsed both old and new with regexes and matched the individual parts

**Review by ScottSuarez — APPROVED** [review]

**Comment by ScottSuarez** [comment]
Looks good to me. Just waiting on tests.

**Pull request merged**